### PR TITLE
general: webui package

### PR DESCRIPTION
**Pull request by tupakkatapa**
Packaging the Web UI in this pull request.

### CODE REVIEW

File: yarn.nix
Comment (jhvst) [inline yarn.nix:1]: Why is this deleted? Don't we need it for building the backend?

File: flake.nix [hunk 1]
```diff
@@ -134,14 +137,14 @@
             ];
           };
 
+          webui = pkgs.callPackage ./webui { };
```
Comment (jhvst) [inline flake.nix:141]: Can we create a directory called `packages` where we have our packages? Like in my nix-config.

File: app.js [hunk 2]
```diff
@@ -7,7 +7,7 @@ const app = express();
 app.use(express.json());
 app.use(cors());
 
-app.use(express.static("webui/dist"));
+app.use(express.static("result/dist"));
```
Comment (jhvst) [inline app.js:10]: Use `substituteInPlace` in favor of hard-coding the result.

File: flake.nix [hunk 3]
```diff
@@ -66,8 +66,11 @@
           server = {
             description = "Initialize and launch the web server";
             exec = ''
-              nix run --no-warn-dirty .#update-json \
-              && nix run .#
+              export NIX_CONFIG='warn-dirty = false' \
+              && nix eval --json .#schema | jq > webui/public/schema.json \
+              && nix run .#update-json \
+              && nix build .#webui \
```
Comment (jhvst) [inline flake.nix:73]: Instead of the line 72, we should have it as a dependency of .#homestakeros.

File: flake.nix [hunk 4]
```diff
@@ -134,14 +137,14 @@
...
+          webui = pkgs.callPackage ./webui { };
           homestakeros = pkgs.mkYarnPackage {
             pname = "homestakeros";
             version = "0.0.1";
 
             src = ./.;
             packageJSON = ./package.json;
             yarnLock = ./yarn.lock;
-            yarnNix = ./yarn.nix;
           };
```
Comment (jhvst) [inline flake.nix:149]: Here, we should patch the code after `unpackPhase`, in a phase called `https://ryantm.github.io/nixpkgs/stdenv/stdenv/#var-stdenv-postUnpack` or `postPatch`.

So:

```
postUnpack = ''
  substituteInPlace ./app.js \
    --replace webui/dist ${webui}/dist
''
```

For example, see: https://github.com/jhvst/nix-config/blob/8c5227fb6e71d4efae8fb60e9a601c5f4aa1d58e/packages/vkroots/default.nix#L22

This way, the webui will become a dependency of the homestakeros package, and the dist will point to nix store path of the webui.

If we do this, then the schema fill will probably have to be included in git, unless it can be called from the build script of the webui.

File: flake.nix [hunk 5]
```diff
@@ -66,8 +66,11 @@
           server = {
             description = "Initialize and launch the web server";
             exec = ''
-              nix run --no-warn-dirty .#update-json \
-              && nix run .#
+              export NIX_CONFIG='warn-dirty = false' \
```
Comment (jhvst) [inline flake.nix:69]: Why is this export defined?

**Review by jhvst — CHANGES_REQUESTED** [review]

File: yarn.nix
Comment (tupakkatapa) [reply yarn.nix:1]: It is optional and generated on the fly. 

Ref: https://github.com/nix-community/yarn2nix#example-defaultnix

File: flake.nix (hunk 5 above)
Comment (tupakkatapa) [reply flake.nix:69]: I did not bother to add `--no-warn-dirty` for all the Nix commands, so I just threw that in. The best solution for this would be to get the `NIX_CONFIG` environment variable defined for the devShell. I tried and failed.

File: yarn.nix
Comment (jhvst) [reply yarn.nix:1]: TIL, sounds good to me then.

File: packages/frontend/flake.nix [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        config,
+        self',
+        inputs',
+        system,
+        ...
+      }: {
+        formatter = nixpkgs.legacyPackages.${system}.alejandra;
+
+        packages.webui = let
+          src = ./.;
```
Comment (jhvst) [inline packages/frontend/flake.nix:39]: webui should be in project root /webui folder (where it was before) and the source should just import it from there

File: packages/frontend/flake.nix [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+          };
+        in
+          pkgs.stdenv.mkDerivation {
+            inherit src version name yarnOfflineCache;
+
+            buildInputs = [
+              pkgs.nodejs_18
+              pkgs.yarn
+              pkgs.yarn2nix-moretea.fixup_yarn_lock
+              inputs'.backend.packages.default
```
Comment (jhvst) [inline packages/frontend/flake.nix:55]: remove this line and the input. the package is not used for anything.

File: packages/backend/app.js [hunk 8]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env node
 import express from "express";
-import apiRouter from "./webui/api.js";
+import apiRouter from "../../webui/api.js";
```
Comment (jhvst) [inline packages/backend/app.js:3]: The api.js should be okay to move in this folder without it causing side-effects on the frontend package.

File: packages/backend/app.js [hunk 9]
```diff
@@ -9,6 +9,6 @@ app.use(cors());
 
 app.use(express.static("webui/dist"));
 app.use("/api", apiRouter);
-app.use("/nixosConfigurations", express.static("./webui/nixosConfigurations"));
+app.use("/nixosConfigurations", express.static("../../webui/nixosConfigurations"));
```
Comment (jhvst) [inline packages/backend/app.js:12]: Notice that this has to be tried after compiling the frontend. When you talk about the relative path, notice that once you have compiled the binary, then the current path might be the flake root path, for which this change might actually break things.

Does it? I don't know, it depends on how nodejs interprets relative paths in this case. Another solution, if necessary, is to use the current working directory path here. Let me know if that is necessary, it will need some additional parameters to be brought into scope.

File: flake.nix [hunk 10]
```diff
@@ -134,16 +144,9 @@
...
-            version = "0.0.1";
-
-            src = ./.;
-            packageJSON = ./package.json;
-            yarnLock = ./yarn.lock;
-            yarnNix = ./yarn.nix;
-          };
-          default = packages.homestakeros;
+          frontend = inputs'.frontend.packages.default;
+          backend = inputs'.backend.packages.default;
```
Comment (jhvst) [inline flake.nix:148]: packages.${pkgs.system}.default

File: packages/backend/flake.nix [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    nixpkgs,
+    flake-parts,
+    frontend,
+    ...
+  }:
+    flake-parts.lib.mkFlake {inherit inputs;} {
+      systems = [
+        #"aarch64-darwin"
+        #"aarch64-linux"
+        #"x86_64-darwin"
```
Comment (jhvst) [inline packages/backend/flake.nix:20]: These comments should be removed -- and can be removed if the system variable is used in the flake root file.

File: flake.nix (hunk 10 above)
Comment (tupakkatapa) [reply flake.nix:147]: Does not need to be clarified when using `inputs'` or `self'`. Ref: https://flake.parts/module-arguments#inputs

**Review by jhvst — APPROVED** [review]

File: flake.nix [hunk 12]
```diff
@@ -15,10 +15,11 @@
...
   }:
     flake-parts.lib.mkFlake {inherit inputs;} {
+      # Listed platforms must be available to the current machine for evaluating the Web UI packages
       systems = [
-        "aarch64-darwin"
-        "aarch64-linux"
-        "x86_64-darwin"
+        # "aarch64-darwin"
+        # "aarch64-linux"
+        # "x86_64-darwin"
```
Comment (jhvst) [inline flake.nix:22]: What is the reason for this?

Comment (tupakkatapa) [reply flake.nix:22]: As the comment a couple of lines above explains, the Web UI packages contain the 'yarn2nix' package, which requires the listed systems to be available on the local machine during evaluation for some reason. Even if we had machines for remote building these kinds of packages, we can't assume that everyone has access to each platform.

Do you have some kind of solution to this? This is the current output with all the platforms listed:
```shell
[I] kari@torque ~/W/HomestakerOS (feat/pkgs-webui)> nix flake check
warning: Git tree '/home/kari/Workspace/HomestakerOS' is dirty
warning: unknown flake output 'schema'
Failed to find a machine for remote build!
derivation: fw0007wxjd7yqjjahv7zj4ng5z7z5hkx-yarn2nix-1.0.0.drv
required (system, features): (aarch64-darwin, [])
2 available machines:
(systems, maxjobs, supportedFeatures, mandatoryFeatures)
([aarch64-linux, i686-linux, x86_64-linux], 20, [benchmark, big-parallel, kvm, nixos-test], [])
([aarch64-linux, armv7l-linux], 16, [benchmark, big-parallel, gccarch-armv8-a, kvm, nixos-test], [])
error:
       … while checking flake output 'packages'

         at «none»:0: (source not available)

       … while checking the derivation 'packages.aarch64-darwin.default'

         at «none»:0: (source not available)

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: a 'aarch64-darwin' with features {} is required to build '/nix/store/fw0007wxjd7yqjjahv7zj4ng5z7z5hkx-yarn2nix-1.0.0.drv', but I am a 'x86_64-linux' with features {benchmark, big-parallel, kvm, nixos-test}
```

Comment (tupakkatapa) [reply flake.nix:22]: > I think it's fine to merge first

I will merge this after I solve the problem with the after build generated JSON data not being picked up by the frontend. Thats a pretty huge problem too.

> Does the output change if it's the systems is flakeExposed?

I tried this and if i recall correctly, it didnt change a thing.

**Comment by tupakkatapa** [comment]
This attempt failed. Closing and removing the branch.
